ludwig/modules/optimization_modules.py

```diff
@@ -12,34 +12,319 @@
 
+def register_optimizer(name: str):
+    def wrap(optimizer_config: BaseOptimizerConfig):
+        optimizer_registry[name] = (optimizer_config.optimizer_class, optimizer_config)
+        return optimizer_config
+
+    return wrap
+
+
+@dataclass
+class BaseOptimizerConfig(BaseMarshmallowConfig, ABC):
+    """Base class for optimizers. Not meant to be used directly.
+
+    The dataclass format prevents arbitrary properties from being set. Consequently, in child classes, all properties
+    from the corresponding `torch.optim.Optimizer` class are copied over: check each class to check which attributes are
+    different from the torch-specified defaults.
+    """
+
+    optimizer_class: ClassVar[Optional[torch.optim.Optimizer]] = None
+    "Class variable pointing to the corresponding `torch.optim.Optimizer` class."
+    type: str
+    """Name corresponding to an optimizer `ludwig.modules.optimization_modules.optimizer_registry`.
+       Technically mutable, but attempting to load a derived optimizer with `type` set to a mismatched value will
+       result in a `ValidationError`."""
+
+
+@register_optimizer(name="sgd")
+@register_optimizer(name="gd")
+@register_optimizer(name="stochastic_gradient_descent")
+@register_optimizer(name="gradient_descent")
+@dataclass
+class SGDOptimizerConfig(BaseOptimizerConfig):
+    """Parameters for stochastic gradient descent."""
+
+    optimizer_class: ClassVar[torch.optim.Optimizer] = torch.optim.SGD
+    """Points to `torch.optim.SGD`."""
+
+    type: str = StringOptions(
+        ["sgd", "gd", "stochastic_gradient_descent", "gradient_descent"], default="sgd", nullable=False
+    )
+    """Must be one of ['sgd', 'gd', 'stochastic_gradient_descent', 'gradient_descent']  - corresponds to names
+       in `ludwig.modules.optimization_modules.optimizer_registry` (default: 'sgd')"""
+
+    lr: float = FloatRange(default=0.001, min=0.0, max=1.0)
+    "(default: 0.001)"
+
+    # Defaults taken from https://pytorch.org/docs/stable/generated/torch.optim.SGD.html#torch.optim.SGD :
+    momentum: float = NonNegativeFloat(default=0.0)
+    weight_decay: float = NonNegativeFloat(default=0.0)
+    dampening: float = NonNegativeFloat(default=0.0)
+    nesterov: bool = False
+
+
+@register_optimizer(name="adam")
+@dataclass
+class AdamOptimizerConfig(BaseOptimizerConfig):
+    """Parameters for adam optimization."""
+
+    optimizer_class: ClassVar[torch.optim.Optimizer] = torch.optim.Adam
+    """Points to `torch.optim.Adam`."""
+
+    type: str = StringOptions(["adam"], default="adam", nullable=False)
+    """Must be 'adam' - corresponds to name in `ludwig.modules.optimization_modules.optimizer_registry`
+       (default: 'adam')"""
+
+    lr: float = FloatRange(default=0.001, min=0.0, max=1.0)
+    """(default: 0.001)"""
+
+    betas: Tuple[float, float] = FloatRangeTupleDataclassField(default=(0.9, 0.999))
+    """(default: (0.9, 0.999))"""
+
+    eps: float = NonNegativeFloat(default=1e-08)
+    """(default: 1e-08)"""
```

Comment:
It looks like these are torch defaults: https://pytorch.org/docs/stable/generated/torch.optim.Adam.html

Reply:
also changed the defaults for the other optimizers, as per offline disc.